First endpoints

patryklason opened the pull request:
1. Added database scheme
2. Created database connection
3. Extracted functions and statements to dedicated files in dedicated packages
4. Created User and UserView models
     - User model contains all user information
     - UserView model contains all user information **except the password**

5. Created first 2 endpoints:

    - get user by email (using User model)
    - get users list by teamId (using UserView model)

6. Added Refined data types to models
    - it turns out that there are some problems related to Refined types (described in the comment in _com.ocadotechnology.models.User_)

    
7. Added some dependencies to build.sbt:
    - "tapir-refined"
    - "tapir-cats"

patryklason committed:
Working getUser endpoint

patryklason committed:
Extract functions to dedicated objects in packages

patryklason committed:
Add corresponding to database value "leadingTeam" to User model

patryklason committed:
Add refined types

patryklason committed:
Add safe UserView model + get safeUsersList endpoint

patryklason committed:
Extract versions to dedicated object in build.sbt

patryklason committed:
Add database scheme

majk-p commented on src/main/scala/com/ocadotechnology/Endpoints.scala:
```diff
@@ -1,42 +1,30 @@
+import com.ocadotechnology.repositories.UserRepository
+import com.ocadotechnology.models.{User, UserView}
+import sttp.tapir.codec.refined.TapirCodecRefined
+
```
```suggestion
import sttp.tapir.codec.refined.*
```
to solve the problems with refined :wink:

majk-p commented on src/main/scala/com/ocadotechnology/models/User.scala:
```diff
@@ -0,0 +1,90 @@
+type Email = String Refined MatchesRegex["""[a-z0-9]+@[a-z0-9]+\\.[a-z0-9]{2,}"""]
+
+/*
+Some unexpected behaviour spotted: when building the project, it sometimes produces error:
+
+
+Could not find Schema for type String Refined
+  eu.timepit.refined.string.MatchesRegex[
+    ("[a-z0-9]+@[a-z0-9]+\\\\.[a-z0-9]{2,}" : String)
+  ].
+Schemas can be derived automatically by adding: `import sttp.tapir.generic.auto._`, or manually using `Schema.derived[T]`.
+The latter is also useful for debugging derivation errors.
+You can find more details in the docs: https://tapir.softwaremill.com/en/latest/endpoint/schemas.html
+When integrating with a third-party datatype library remember to import respective schemas/codecs as described in https://tapir.softwaremill.com/en/latest/endpoint/integrations.html.
+I found:
+
+    sttp.tapir.Schema.derivedEnumerationValue[Enumeration#Value]
+
+But method derivedEnumerationValue in trait SchemaCompanionMacros does not match type sttp.tapir.Schema[String Refined
+  eu.timepit.refined.string.MatchesRegex[
+    ("[a-z0-9]+@[a-z0-9]+\\\\.[a-z0-9]{2,}" : String)
+  ]
+].
+
+One of the following imports might make progress towards fixing the problem:
+
+  import shapeless.~?>.idKeyWitness
+  import shapeless.~?>.idValueWitness
+  import shapeless.~?>.witness
+
+
+    .out(jsonBody[User])
+
+
+
+
+The only working solution I found is to switch back to generic types, build the project and then switch to refined types and build the project again
+Turns out that every change in Endpoints.scala (even adding a new line) results in throwing this error.
+*/
```
> The only working solution I found is to switch back to generic types, build the project and then switch to refined types and build the project again

So at this point you basically rely on the binary version of the file, as long as it was compiled once, and the data type was valid at that point it would work on partial compilation I suppose :wink:  Nevertheless I appreciate the effort, it's an interesting finding on its own!

The solution here is very simple, just accept the suggestion at https://github.com/ocadotechnology-university/doubloons/pull/2/files#r1148822552 

It's actually in the error message, it says:

> When integrating with a third-party datatype library remember to import respective schemas/codecs as described in https://tapir.softwaremill.com/en/latest/endpoint/integrations.html.

Now of course this might not sound obvious at the first sight, but here we rely on the tapir-refined integration, it's in the `build.sbt` file
```scala
"com.softwaremill.sttp.tapir" %% "tapir-refined" % Versions.tapir,
```
In the [datatype integrations](https://tapir.softwaremill.com/en/latest/endpoint/integrations.html) document mentioned in the error, you'll find the following section about refined:

![tapir-integration-docs](https://user-images.githubusercontent.com/7260594/227855618-306f8ba7-4f1a-4d5e-8cf5-c1bfa9a37b3e.png)

So the solution is to simply add the import:
```scala
import sttp.tapir.codec.refined.*
```

This is required because Tapir generates implicit `Schema` objects, that represents a bunch of metadata about the endpoint, including the codecs and endpoint validation rules. When all goes well, you shouldn't have to worry about it, and the endpoint simply summons some extra implicits to the context to make sure Tapir is able to generate schemas that include refined types.

You can learn more about them in https://tapir.softwaremill.com/en/latest/endpoint/codecs.html#schemas

patryklason replied:
You are right. I added the tapir-refined module to the dependencies, but didn't manage to use the correct import statement.
Thank You for the feedback!

patryklason committed:
Update src/main/scala/com/ocadotechnology/Endpoints.scala

Co-authored-by: Michał Pawlik <majk-p@users.noreply.github.com>

majk-p commented on build.sbt:
```diff
@@ -6,6 +6,10 @@ val Versions = new {
+  val timepit = "0.10.2"
+  val tapirDoobie = "0.19.0-M7"
+  val postgresql = "42.5.4"
+  val doobie = "1.0.0-RC1"
```
Can be updated to RC2
```suggestion
  val doobie = "1.0.0-RC2"
```

majk-p commented on build.sbt:
```diff
@@ -6,6 +6,10 @@ val Versions = new {
   val circe = "3.8.13"
   val scalaTest = "3.2.15"
+  val timepit = "0.10.2"
+  val tapirDoobie = "0.19.0-M7"
```
looks unused :wink:

patryklason committed:
Optimize imports

patryklason committed:
Update build.sbt

patryklason committed:
Refactor: add traits for Services and Repositories

patryklason committed:
Add create-user endpoint

patryklason committed:
Change "getUserByEmail" endpoint to use UserView model

patryklason committed:
Refactor: rename vals

majk-p commented on build.sbt:
```diff
@@ -22,7 +24,15 @@ lazy val rootProject = (project in file(".")).settings(
       "com.softwaremill.sttp.tapir" %% "tapir-sttp-stub-server" % Versions.tapir % Test,
       "org.scalatest" %% "scalatest" % Versions.scalaTest % Test,
-      "com.softwaremill.sttp.client3" %% "circe" % Versions.circe % Test
+      "com.softwaremill.sttp.client3" %% "circe" % Versions.circe % Test,
```
I'd rename `Versions.circe` to `Versions.sttpCirce` since this is not the version of Circe itself but the sttp-circe integration.

majk-p commented on build.sbt:
```diff
@@ -22,7 +24,15 @@ lazy val rootProject = (project in file(".")).settings(
       "org.scalatest" %% "scalatest" % Versions.scalaTest % Test,
-      "com.softwaremill.sttp.client3" %% "circe" % Versions.circe % Test
+      "com.softwaremill.sttp.client3" %% "circe" % Versions.circe % Test,
+      "io.circe" %% "circe-refined" % "0.14.5",
```
Here is the real version of the circe library (https://github.com/circe/circe). This version can be extracted ad `Versions.circe`

majk-p commented on build.sbt:
```diff
@@ -22,7 +24,15 @@ lazy val rootProject = (project in file(".")).settings(
+      "org.tpolecat" %% "doobie-core" % Versions.doobie,
+      "org.tpolecat" %% "doobie-refined" % Versions.doobie,
+      "org.tpolecat" %% "doobie-postgres" % Versions.doobie,
+      "org.tpolecat" %% "doobie-scalatest" % Versions.doobie % "test",
```
To make things uniform, I think this is the preferred approach
```suggestion
      "org.tpolecat" %% "doobie-scalatest" % Versions.doobie % Test,
```

majk-p commented on src/main/scala/com/ocadotechnology/Endpoints.scala:
```diff
@@ -1,42 +1,33 @@
+    .in("user" / path[String]("email").example("admin@example.com"))
+    .description("TODO")
+    .out(jsonBody[UserView])
+    .errorOut(jsonBody[String])
```
In the next iteration we can work toward making those endpoints already validate the data and take `Email` as an input.

majk-p commented on src/main/scala/com/ocadotechnology/models/User.scala:
```diff
@@ -0,0 +1,27 @@
+                  password: NonEmptyString,
+                  avatar: Option[NonEmptyString],
+                  leadingTeam: Option[Boolean],
+                )
```
In the next iteration let's change the member types to Value Classes (https://docs.scala-lang.org/overviews/core/value-classes.html)

majk-p commented on src/main/scala/com/ocadotechnology/repositories/UserViewRepository.scala:
```diff
@@ -0,0 +1,35 @@
+
+/**
+ * Implements CRUD operations on data entities for UserView model
+ */
```
Technically CRUD stands for Create Read Update Delete and this one is intended only to do the Read part :wink:

majk-p commented on src/main/scala/com/ocadotechnology/services/UserService.scala:
```diff
@@ -0,0 +1,24 @@
+ */
+
+trait UserService {
+  def createUserLogic(user: User): IO[Either[String, String]]
```
I think I'd change this to `createUser` for simplicity. Also the return type could be improved. If it returns a meaningful information on success, let's use a value class to indicate that (return user Id or something).

The user creation process shouldn't create any reasonable information on the output since we don't generate any IDs upon user creation, so perhaps we are good with returning `IO[Either[String, Unit]]`?

majk-p commented on src/main/scala/com/ocadotechnology/services/UserViewService.scala:
```diff
@@ -0,0 +1,39 @@
+
+trait UserViewService {
+  def getUserByEmailLogic(email: String): IO[Either[String, UserView]]
+  def getUsersByTeamIdLogic(teamId: String): IO[Either[String, List[UserView]]]
```
I think everything in this trait would end with `Logic` so we might as well remove it from the method names for conciseness

majk-p reviewed (approved):
Looks good, few minor things to do now and some changes for the next iterations.

patryklason committed:
Update build.sbt

patryklason committed:
Update 'create-user' endpoint, Rename methods

ressenov committed:
Updated database schema

1. Dates are all varchars(7) in such format: "04-2023"
2. Added table team_leaders
3. Adjusted table comments to support only 1 comment for all doubloons given to a user
4. Refactoring of naming convention - migrating to snake case

patryklason committed:
Update models and repositories to match new database schema

pkowalcze commented on src/test/scala/com/ocadotechnology/EndpointsSpec.scala:
```diff
@@ -17,7 +16,7 @@ import sttp.tapir.integ.cats.CatsMonadError
 class EndpointsSpec extends AnyFlatSpec with Matchers with EitherValues:
 
-  it should "return hello message" in {
+  /*it should "return hello message" in {
```
It would be good to replace this with actual tests

pkowalcze commented on src/main/scala/com/ocadotechnology/models/User.scala:
```diff
@@ -0,0 +1,26 @@
+                  avatar: Option[NonEmptyString],
+                )
+object User {
+  implicit val codec: Codec[User] = deriveCodec
```
You could use Scala 3's anonymous `given` like this:

```scala
given Codec[User] = deriveCodec
```

pkowalcze commented on src/main/scala/com/ocadotechnology/models/User.scala:
```diff
@@ -0,0 +1,26 @@
+
+case class User (
+                  email: Email,
+                  teamId: Option[PosInt],
```
Let's change `teamId` to `NonEmptyString`, so that it's easier to migrate to UUID in the future

patryklason committed:
Change teamId type: PosInt -> NonEmptyString

patryklason committed:
Update db-scheme.sql: team_id integer -> character varying

patryklason committed:
Update models codecs declaration to use Scala 3's `given`

ressenov committed:
1. Add case classes for doubloons and comments
2. Add endpoints for doubloons
3. Refactor packages to implement sorting by abstraction
4. Add api/ to all urls

patryklason committed:
Add: getCurrentSpentDoubloons endpoint

patryklason committed:
Refactor: standardize code with braces

patryklason committed:
Add: `getCurrentComments`, `addComent`, `updateComment`, `deleteComment` endpoints

patryklason committed:
Refactor: standardize api urls
